### ISSUE TITLE
Any tricks necessary to use with Istio mTLS?

### ISSUE THREAD

**Issue by funkypenguin** [zalando/postgres-operator#772]
Hi folks,

Now that I've successfully been able to launch postgres-operator v1.3.0, I've tried to deploy the minimal manifest, and run into another problem. I expect it's related to the fact that we're using Istio sidecar injection for mTLS.

Here's my postgres-operator logs:

```
time="2019-12-18T08:39:26Z" level=info msg="\"ADD\" event has been queued" cluster-name=dev/acid-minimal-cluster pkg=controller worker=0
time="2019-12-18T08:39:26Z" level=info msg="creation of the cluster started" cluster-name=dev/acid-minimal-cluster pkg=controller worker=0
time="2019-12-18T08:39:26Z" level=info msg="endpoint \"dev/acid-minimal-cluster\" has been successfully created" cluster-name=dev/acid-minimal-cluster pkg=cluster worker=0
time="2019-12-18T08:39:26Z" level=info msg="master service \"dev/acid-minimal-cluster\" has been successfully created" cluster-name=dev/acid-minimal-cluster pkg=cluster worker=0
time="2019-12-18T08:39:26Z" level=debug msg="No load balancer created for the replica service" cluster-name=dev/acid-minimal-cluster pkg=cluster worker=0
time="2019-12-18T08:39:26Z" level=info msg="replica service \"dev/acid-minimal-cluster-repl\" has been successfully created" cluster-name=dev/acid-minimal-cluster pkg=cluster worker=0
time="2019-12-18T08:39:26Z" level=debug msg="team API is disabled, returning empty list of members for team \"acid\"" cluster-name=dev/acid-minimal-cluster pkg=cluster worker=0
time="2019-12-18T08:39:26Z" level=info msg="users have been initialized" cluster-name=dev/acid-minimal-cluster pkg=cluster worker=0
time="2019-12-18T08:39:26Z" level=debug msg="secret \"dev/foo-user.acid-minimal-cluster.credentials.postgresql.acid.zalan.do\" already exists, fetching its password" cluster-name=dev/acid-minimal-cluster pkg=cluster worker=0
time="2019-12-18T08:39:26Z" level=debug msg="secret \"dev/zalando.acid-minimal-cluster.credentials.postgresql.acid.zalan.do\" already exists, fetching its password" cluster-name=dev/acid-minimal-cluster pkg=cluster worker=0
time="2019-12-18T08:39:27Z" level=debug msg="secret \"dev/postgres.acid-minimal-cluster.credentials.postgresql.acid.zalan.do\" already exists, fetching its password" cluster-name=dev/acid-minimal-cluster pkg=cluster worker=0
time="2019-12-18T08:39:27Z" level=debug msg="secret \"dev/standby.acid-minimal-cluster.credentials.postgresql.acid.zalan.do\" already exists, fetching its password" cluster-name=dev/acid-minimal-cluster pkg=cluster worker=0
time="2019-12-18T08:39:27Z" level=info msg="secrets have been successfully created" cluster-name=dev/acid-minimal-cluster pkg=cluster worker=0
time="2019-12-18T08:39:27Z" level=info msg="pod disruption budget \"dev/postgres-acid-minimal-cluster-pdb\" has been successfully created" cluster-name=dev/acid-minimal-cluster pkg=cluster worker=0
time="2019-12-18T08:39:27Z" level=debug msg="Generating Spilo container, environment variables: [{SCOPE acid-minimal-cluster nil} {PGROOT /home/postgres/pgdata/pgroot nil} {POD_IP  &EnvVarSource{FieldRef:&ObjectFieldSelector{APIVersion:v1,FieldPath:status.podIP,},ResourceFieldRef:nil,ConfigMapKeyRef:nil,SecretKeyRef:nil,}} {POD_NAMESPACE  &EnvVarSource{FieldRef:&ObjectFieldSelector{APIVersion:v1,FieldPath:metadata.namespace,},ResourceFieldRef:nil,ConfigMapKeyRef:nil,SecretKeyRef:nil,}} {PGUSER_SUPERUSER postgres nil} {KUBERNETES_SCOPE_LABEL cluster-name nil} {KUBERNETES_ROLE_LABEL spilo-role nil} {KUBERNETES_LABELS application=spilo nil} {PGPASSWORD_SUPERUSER  &EnvVarSource{FieldRef:nil,ResourceFieldRef:nil,ConfigMapKeyRef:nil,SecretKeyRef:&SecretKeySelector{LocalObjectReference:LocalObjectReference{Name:postgres.acid-minimal-cluster.credentials.postgresql.acid.zalan.do,},Key:password,Optional:nil,},}} {PGUSER_STANDBY standby nil} {PGPASSWORD_STANDBY  &EnvVarSource{FieldRef:nil,ResourceFieldRef:nil,ConfigMapKeyRef:nil,SecretKeyRef:&SecretKeySelector{LocalObjectReference:LocalObjectReference{Name:standby.acid-minimal-cluster.credentials.postgresql.acid.zalan.do,},Key:password,Optional:nil,},}} {PAM_OAUTH2  nil} {HUMAN_ROLE zalandos nil} {SPILO_CONFIGURATION {\"postgresql\":{\"bin_dir\":\"/usr/lib/postgresql/11/bin\"},\"bootstrap\":{\"initdb\":[{\"auth-host\":\"md5\"},{\"auth-local\":\"trust\"}],\"users\":{\"zalandos\":{\"password\":\"\",\"options\":[\"CREATEDB\",\"NOLOGIN\"]}},\"dcs\":{}}} nil} {DCS_ENABLE_KUBERNETES_API true nil}]" cluster-name=dev/acid-minimal-cluster pkg=cluster worker=0
time="2019-12-18T08:39:27Z" level=debug msg="created new statefulset \"dev/acid-minimal-cluster\", uid: \"dbdbb202-5ce0-45b0-867f-8a52712d850d\"" cluster-name=dev/acid-minimal-cluster pkg=cluster worker=0
time="2019-12-18T08:39:27Z" level=info msg="statefulset \"dev/acid-minimal-cluster\" has been successfully created" cluster-name=dev/acid-minimal-cluster pkg=cluster worker=0
time="2019-12-18T08:39:27Z" level=info msg="waiting for the cluster being ready" cluster-name=dev/acid-minimal-cluster pkg=cluster worker=0

time="2019-12-18T08:39:42Z" level=debug msg="Waiting for 2 pods to become ready" cluster-name=dev/acid-minimal-cluster pkg=cluster worker=0
```

The acid-minimal-cluster-0 pod starts up normally, initializes itself, and then logs the following regularly (which I assume is good):

```
2019-12-18 09:18:04.507 37 LOG {ticks: 0, maint: 0, retry: 0}
2019-12-18 09:18:10,122 INFO: Lock owner: acid-minimal-cluster-0; I am acid-minimal-cluster-0
2019-12-18 09:18:10,139 INFO: no action.  i am the leader with the lock
```

But my acid-minimal-cluster-1 pod logs this:

```
root@cn1:~# kubectl logs -n dev acid-minimal-cluster-1 postgres
decompressing spilo image...
2019-12-18 08:40:23,105 - bootstrapping - INFO - Figuring out my environment (Google? AWS? Openstack? Local?)
2019-12-18 08:40:25,117 - bootstrapping - INFO - Could not connect to 169.254.169.254, assuming local Docker setup
2019-12-18 08:40:25,121 - bootstrapping - INFO - No meta-data available for this provider
2019-12-18 08:40:25,122 - bootstrapping - INFO - Looks like your running local
2019-12-18 08:40:25,153 - bootstrapping - WARNING - could not parse kubernetes labels as a JSON: Expecting value: line 1 column 1 (char 0), reverting to the default: {"application": "spilo"}
2019-12-18 08:40:25,180 - bootstrapping - INFO - Configuring patroni
2019-12-18 08:40:25,201 - bootstrapping - INFO - Writing to file /home/postgres/postgres.yml
2019-12-18 08:40:25,201 - bootstrapping - INFO - Configuring crontab
2019-12-18 08:40:25,216 - bootstrapping - INFO - Configuring certificate
2019-12-18 08:40:25,216 - bootstrapping - INFO - Generating ssl certificate
2019-12-18 08:40:25,279 - bootstrapping - INFO - Configuring pgbouncer
2019-12-18 08:40:25,280 - bootstrapping - INFO - No PGBOUNCER_CONFIGURATION was specified, skipping
2019-12-18 08:40:25,280 - bootstrapping - INFO - Configuring pam-oauth2
2019-12-18 08:40:25,281 - bootstrapping - INFO - No PAM_OAUTH2 configuration was specified, skipping
2019-12-18 08:40:25,281 - bootstrapping - INFO - Configuring wal-e
2019-12-18 08:40:25,281 - bootstrapping - INFO - Configuring pgqd
2019-12-18 08:40:25,282 - bootstrapping - INFO - Configuring patronictl
2019-12-18 08:40:25,282 - bootstrapping - INFO - Configuring renice
2019-12-18 08:40:25,288 - bootstrapping - INFO - Skipping creation of renice cron job due to lack of permissions
2019-12-18 08:40:25,288 - bootstrapping - INFO - Configuring standby-cluster
2019-12-18 08:40:25,288 - bootstrapping - INFO - Configuring log
2019-12-18 08:40:25,289 - bootstrapping - INFO - Configuring bootstrap
2019-12-18 08:40:27,082 INFO: No PostgreSQL configuration items changed, nothing to reload.
2019-12-18 08:40:27,089 INFO: Lock owner: acid-minimal-cluster-0; I am acid-minimal-cluster-1
2019-12-18 08:40:27,523 INFO: trying to bootstrap from leader 'acid-minimal-cluster-0'
pg_basebackup: error: could not connect to server: received invalid response to SSL negotiation:
2019-12-18 08:40:27,542 ERROR: Error when fetching backup: pg_basebackup exited with code=1
2019-12-18 08:40:27,543 WARNING: Trying again in 5 seconds
pg_basebackup: error: could not connect to server: received invalid response to SSL negotiation:
2019-12-18 08:40:32,563 ERROR: Error when fetching backup: pg_basebackup exited with code=1
2019-12-18 08:40:32,564 ERROR: failed to bootstrap from leader 'acid-minimal-cluster-0'
2019-12-18 08:40:32,565 INFO: Removing data directory: /home/postgres/pgdata/pgroot/data
```

Suspecting that the above indicates that the pods can't talk to each other, I grabbed the istio-proxy sidecar logs, which look like this for acid-minimal-cluster-1:

```
[2019-12-18T09:19:27.109Z] "- - -" 0 - "-" "-" 8 7 2 - "-" "-" "-" "-" "10.120.120.193:5432" PassthroughCluster 10.114.44.181:35974 10.120.120.193:5432 10.114.44.181:35972 - -
[2019-12-18T09:19:32.134Z] "- - -" 0 - "-" "-" 8 7 1 - "-" "-" "-" "-" "10.120.120.193:5432" PassthroughCluster 10.114.44.181:36708 10.120.120.193:5432 10.114.44.181:36706 - -
[2019-12-18T09:19:37.109Z] "- - -" 0 - "-" "-" 8 7 2 - "-" "-" "-" "-" "10.120.120.193:5432" PassthroughCluster 10.114.44.181:37564 10.120.120.193:5432 10.114.44.181:37562 - -
[2019-12-18T09:19:42.132Z] "- - -" 0 - "-" "-" 8 7 1 - "-" "-" "-" "-" "10.120.120.193:5432" PassthroughCluster 10.114.44.181:38270 10.120.120.193:5432 10.114.44.181:38268 - -
```

The **PassthroughCluster** bit is suspicious, since it would indicate that mTLS is _not_ being employed - perhaps because we're relying on the creation of endpoints, which may not make it into the service mesh?

I've noticed some comments in the helm chart around istio labels, but I haven't found anything indicating that certain config is explicitly required for Istio.

Can you point me to some docs / examples with postgres-operator working successfully with Istio?

Thanks!
David

**Comment by FxKu** [comment]
No Istio experience from my side, but check the thread in PR #657. If I understand correctly, with the new custom pod annotations you can prevent Istio from interfering the Spilo pods.

**Comment by funkypenguin** [comment]
Right, for posterity, the solution here has to do with the fact that in order for istio to permit pod-to-pod communication, a ServiceEntry must be created, so that each pod exists as a "service" in the mesh.

I solved the issue by creating the following (explicitly specifying each potential participating pod):

```
apiVersion: networking.istio.io/v1alpha3
kind: ServiceEntry
metadata:
  name: acid-minimal-cluster-service-entry
spec:
  hosts:
  - acid-minimal-cluster-0.retort-database.dev.svc.cluster.local
  - acid-minimal-cluster-1.retort-database.dev.svc.cluster.local
  - acid-minimal-cluster-2.retort-database.dev.svc.cluster.local
  - acid-minimal-cluster-3.retort-database.dev.svc.cluster.local  
  location: MESH_INTERNAL
  ports:
  - number: 5432
    name: postgres
    protocol: TCP
  resolution: NONE
```

**Comment by cortopy** [comment]
@funkypenguin worked for me, but I also have to add a DestinationRule for each node so that the SNI gets added to the endpoint. 

I use Istio with mTLS enabled (STRICT) for the full mesh

Example:

```yaml
apiVersion: networking.istio.io/v1beta1
kind: DestinationRule
metadata:
  name: tcp-cluster-0
spec:
  host: acid-kratos-0.default.svc.cluster.local
  trafficPolicy:
    tls:
      mode: ISTIO_MUTUAL
      sni: acid-kratos-0.default.svc.cluster.local
```

Without the DestinationRule, Envoy will output `filter_chain_not_found` as the sni will be missing

**Comment by cortopy** [comment]
Also, I just found out that the serviceEntry needs to have a port for patroni too, so it'd be something like:

```
apiVersion: networking.istio.io/v1alpha3
kind: ServiceEntry
metadata:
  name: acid-minimal-cluster-service-entry
spec:
  hosts:
  - acid-minimal-cluster-0.acid-minimal-cluster.dev.svc.cluster.local
  - acid-minimal-cluster-1.acid-minimal-cluster.dev.svc.cluster.local
  - acid-minimal-cluster-2.acid-minimal-cluster.dev.svc.cluster.local
  - acid-minimal-cluster-3.acid-minimal-cluster.dev.svc.cluster.local  
  location: MESH_INTERNAL
  ports:
  - number: 5432
    name: tcp-postgres
    protocol: TCP
  - number: 8008
    name: tcp-patroni
    protocol: TCP
  resolution: NONE
```

The pods also have 8080 port but I don't know what that one is for

**Comment by krishnachalla-hv** [comment]
After doing the necessary changes suggested by @funkypenguin inter pod communication is working fine. But when `enableConnectionPooler: true` I am seeing below errors in pgbouncer log
`2021-06-01 15:01:25.698 UTC [1] LOG stats: 0 xacts/s, 0 queries/s, in 0 B/s, out 0 B/s, xact 0 us, query 0 us, wait 0 us
2021-06-01 15:02:11.714 UTC [1] LOG C-0x56242d327a70: (nodb)/(nouser)@127.0.0.1:48248 registered new auto-database: db=testdb
2021-06-01 15:02:11.834 UTC [1] WARNING lookup failed: my-test-pg-cluster: result=0`

**Comment by marcoslarsen** [comment]
Hello everyone. I know that this issue is already closed but I'm still having issues setting this up. On our setup we have istio sidecar injection on the pods like described in the original post, and the issue is  pod to pod communication in this setup when we have multiple instances. I've tried the suggested approach, to create the serviceentry for postgres and patroni tcp ports and  the pods hosts. Apparently works, but we found an issue with this solution. ISTIO service entry for TCP ports requires the ip addresses of the pods to be specified and given their dynamic nature it snot straightforward  to get them (if a pod is recreated we require to update the service entry). If addresses are not specified, the entry acts as a proxy matching only the specified port. What we're experiencing is a service that we have that requires to connect to an external database  on port 5432 and the service entry proxies that to our  database cluster. We've tried to prevent istio sidecar injection in the pods using the custom annotations but that causes our service (which requires the istio sidecar) to not be able to connect to our cluster. I was looking also if it was possible to change the default database port for our cluster (instead o 5432 ) but haven't figured that out yet.... Any suggestions?  (note: prevent istio sidecar injection is not something that we actually want to go for). 
The only solutions I have till now is to assume that the port 5432 is taken by our service and no other service (external in particular) that we might need to access can use that port, which is not feasible, or disable HA by enforcing only one available  instance for our cluster.
